File: rust/src/models/checkout_session.rs
// this file is @generated
use serde::{Deserialize, Serialize};

use super::{
    checkout_session_id::CheckoutSessionId, checkout_session_status::CheckoutSessionStatus,
    checkout_type::CheckoutType, customer_id::CustomerId, plan_version_id::PlanVersionId,
    subscription_id::SubscriptionId,
};

#[derive(Clone, Debug, Default, PartialEq, Deserialize, Serialize)]
pub struct CheckoutSession {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub billing_day_anchor: Option<i32>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub billing_start_date: Option<String>,

    pub checkout_type: CheckoutType,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub checkout_url: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub coupon_code: Option<String>,

    pub created_at: String,

    pub customer_id: CustomerId,

    /// When the session expires. None means the session never expires.
    ///
    /// RFC3339 date string.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,

    pub id: CheckoutSessionId,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub net_terms: Option<i32>,

    pub plan_version_id: PlanVersionId,

    pub status: CheckoutSessionStatus,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub subscription_id: Option<SubscriptionId>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub trial_duration_days: Option<i32>,
}

impl CheckoutSession {
    pub fn new(
        checkout_type: CheckoutType,
        created_at: String,
        customer_id: CustomerId,
        id: CheckoutSessionId,
        plan_version_id: PlanVersionId,
        status: CheckoutSessionStatus,
    ) -> Self {
        Self {
            billing_day_anchor: None,
            billing_start_date: None,
            checkout_type,
            checkout_url: None,
            completed_at: None,
            coupon_code: None,
            created_at,
            customer_id,
            expires_at: None,
            id,
            net_terms: None,
            plan_version_id,
            status,
            subscription_id: None,
            trial_duration_days: None,
        }
    }
}

File: typescript/src/models/checkoutSession.ts
// this file is @generated
import { parseDateTime } from "../datetime";
import { type CheckoutSessionId, CheckoutSessionIdSerializer } from "./checkoutSessionId";
import {
  type CheckoutSessionStatus,
  CheckoutSessionStatusSerializer,
} from "./checkoutSessionStatus";
import { type CheckoutType, CheckoutTypeSerializer } from "./checkoutType";
import { type CustomerId, CustomerIdSerializer } from "./customerId";
import {
  type PaymentMethodsConfig,
  PaymentMethodsConfigSerializer,
} from "./paymentMethodsConfig";
import { type PlanVersionId, PlanVersionIdSerializer } from "./planVersionId";
import { type SubscriptionId, SubscriptionIdSerializer } from "./subscriptionId";

export interface CheckoutSession {
  billingDayAnchor?: number | null;

  billingStartDate?: string | null;

  checkoutType: CheckoutType;

  checkoutUrl?: string | null;

  completedAt?: Date | null;

  couponCode?: string | null;

  createdAt: Date;

  customerId: CustomerId;

  /** When the session expires. None means the session never expires. */
  expiresAt?: Date | null;

  id: CheckoutSessionId;

  netTerms?: number | null;

  paymentMethodsConfig?: PaymentMethodsConfig | null;

  planVersionId: PlanVersionId;

  status: CheckoutSessionStatus;

  subscriptionId?: SubscriptionId | null;

  trialDurationDays?: number | null;
}

export const CheckoutSessionSerializer = {
  _fromJsonObject(object: any): CheckoutSession {
    return {
      billingDayAnchor: object["billing_day_anchor"],
      billingStartDate: object["billing_start_date"],
      checkoutType: CheckoutTypeSerializer._fromJsonObject(object["checkout_type"]),
      checkoutUrl: object["checkout_url"],
      completedAt:
        object["completed_at"] != null
          ? parseDateTime(object["completed_at"])
          : undefined,
      couponCode: object["coupon_code"],
      createdAt: parseDateTime(object["created_at"]),
      customerId: CustomerIdSerializer._fromJsonObject(object["customer_id"]),
      expiresAt:
        object["expires_at"] != null ? parseDateTime(object["expires_at"]) : undefined,
      id: CheckoutSessionIdSerializer._fromJsonObject(object["id"]),
      netTerms: object["net_terms"],
      paymentMethodsConfig:
        object["payment_methods_config"] != null
          ? PaymentMethodsConfigSerializer._fromJsonObject(
              object["payment_methods_config"]
            )
          : undefined,
      planVersionId: PlanVersionIdSerializer._fromJsonObject(object["plan_version_id"]),
      status: CheckoutSessionStatusSerializer._fromJsonObject(object["status"]),
      subscriptionId:
        object["subscription_id"] != null
          ? SubscriptionIdSerializer._fromJsonObject(object["subscription_id"])
          : undefined,
      trialDurationDays: object["trial_duration_days"],
    };
  },

  _toJsonObject(self: CheckoutSession): any {
    return {
      billing_day_anchor: self.billingDayAnchor,
      billing_start_date: self.billingStartDate,
      checkout_type: CheckoutTypeSerializer._toJsonObject(self.checkoutType),
      checkout_url: self.checkoutUrl,
      completed_at: self.completedAt,
      coupon_code: self.couponCode,
      created_at: self.createdAt,
      customer_id: CustomerIdSerializer._toJsonObject(self.customerId),
      expires_at: self.expiresAt,
      id: CheckoutSessionIdSerializer._toJsonObject(self.id),
      net_terms: self.netTerms,
      payment_methods_config:
        self.paymentMethodsConfig != null
          ? PaymentMethodsConfigSerializer._toJsonObject(self.paymentMethodsConfig)
          : undefined,
      plan_version_id: PlanVersionIdSerializer._toJsonObject(self.planVersionId),
      status: CheckoutSessionStatusSerializer._toJsonObject(self.status),
      subscription_id:
        self.subscriptionId != null
          ? SubscriptionIdSerializer._toJsonObject(self.subscriptionId)
          : undefined,
      trial_duration_days: self.trialDurationDays,
    };
  },
};
